packages/dd-trace/src/llmobs/sdk.js

```diff
@@ -0,0 +1,448 @@
+'use strict'
+
+const { SPAN_KIND, OUTPUT_VALUE } = require('./constants')
+
+const {
+  validKind,
+  getName,
+  getFunctionArguments
+} = require('./util')
+const { isTrue } = require('../util')
+
+// storage - context management
+const { storage } = require('../../../datadog-core')
+
+const Span = require('../opentracing/span')
+
+const tracerVersion = require('../../../../package.json').version
+const logger = require('../log')
+
+const LLMObsTagger = require('./tagger')
+
+// communicating with writer
+const { flushCh, evalMetricAppendCh } = require('./channels')
+const NoopLLMObs = require('./noop')
+
+class LLMObs extends NoopLLMObs {
+  constructor (tracer, llmobsModule, config) {
+    super(tracer)
+
+    this._config = config
+    this._llmobsModule = llmobsModule
+    this._tagger = new LLMObsTagger(config)
+  }
+
+  get enabled () {
+    return this._config.llmobs.enabled
+  }
+
+  enable (options = {}) {
+    if (this.enabled) {
+      logger.debug('LLMObs is already enabled.')
+      return
+    }
+
+    logger.debug('Enabling LLMObs')
+
+    const { mlApp, agentlessEnabled, apiKey } = options
+
+    const { DD_LLMOBS_ENABLED } = process.env
+
+    const llmobsConfig = {
+      mlApp,
+      agentlessEnabled,
+      apiKey
+    }
+
+    const enabled = DD_LLMOBS_ENABLED == null || isTrue(DD_LLMOBS_ENABLED)
+    if (!enabled) {
+      logger.debug('LLMObs.enable() called when DD_LLMOBS_ENABLED is false. No action taken.')
+      return
+    }
+
+    this._config.llmobs.enabled = true
+    this._config.configure({ ...this._config, llmobs: llmobsConfig })
+
+    // configure writers and channel subscribers
+    this._llmobsModule.enable(this._config)
+  }
+
+  disable () {
+    if (!this.enabled) {
+      logger.debug('LLMObs is already disabled.')
+      return
+    }
+
+    logger.debug('Disabling LLMObs')
+
+    this._config.llmobs.enabled = false
+
+    // disable writers and channel subscribers
+    this._llmobsModule.disable()
+  }
+
+  trace (options = {}, fn) {
+    if (!this.enabled) {
+      logger.warn('Span started while LLMObs is disabled. Spans will not be sent to LLM Observability.')
+    }
+
+    if (typeof options === 'function') {
+      fn = options
+      options = {}
+    }
+
+    const kind = options.kind
+    const valid = validKind(kind)
+    if (!valid) {
+      logger.warn(`Invalid span kind specified: ${kind}. Span will not be sent to LLM Observability.`)
+    }
+
+    const name = getName(kind, options)
+
+    const {
+      spanOptions,
+      ...llmobsOptions
+    } = this._extractOptions(options)
+
+    if (fn.length > 1) {
```

Comment:
Just to clarify here - `trace` works as both an inline span generator and a function wrapper right? what is `fn.length > 1` checking for here?

Reply:
trace only works with an inline anonymous function/block, not wrapping. `fn.length` here returns the number of arguments a function takes, so we're checking if the anonymous function we're tracing takes a callback as well as a span. if it takes a callback, we finish when the callback is called. this is [already handled](https://github.com/DataDog/dd-trace-js/blob/49c20e8d07c5767173a3e39088e2180cc1e62c49/packages/dd-trace/src/tracer.js#L71-L102) by the underlying `tracer.trace`, but I had to redefine it since we need to restore our llmobs span context. lmk if that doesn't make sense!

ref:

```javascript
// finishes when the anonymous function returns
llmobs.trace({ kind: ..., name: ...} () => {})
llmobs.trace({ kind: ..., name: ... }, span => {}) 

llmobs.trace({ kind ..., name ...}, (span, cb) => {
  cb() // finishes here, which could be part of another promise or function that takes a callback
})